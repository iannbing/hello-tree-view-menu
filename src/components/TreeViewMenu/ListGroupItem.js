import React from 'react';
import styled from 'react-emotion';
import { ListGroupItem as DefaultListGroupItem } from 'reactstrap';

const DEFAULT_PADDING = 1.25;
const LEVEL_SPACE = 1;

const ToggleIcon = styled('div')({
  position: 'absolute',
  left: 20,
  width: 10,
  height: 10
});

const ListGroupItemWithPadding = styled(DefaultListGroupItem)(
  ({ level = 0 }) => ({
    paddingLeft: `${DEFAULT_PADDING + level * LEVEL_SPACE}rem`
  })
);

const ListGroupItem = ({
  hasSubItems = false,
  isOpen = false,
  level = 0,
  children,
  ...props
}) => (
  <>
    <ListGroupItemWithPadding level={hasSubItems ? level : 0} {...props}>
      {hasSubItems && <ToggleIcon>{isOpen ? '-' : '+'}</ToggleIcon>}
      {children}
    </ListGroupItemWithPadding>
  </>
);

export default ListGroupItem;
