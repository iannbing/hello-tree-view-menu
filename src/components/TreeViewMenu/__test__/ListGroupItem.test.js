import React from 'react';
import { shallow } from 'enzyme';

import ListGroupItem from '../ListGroupItem';

describe('TreeViewMenu', () => {
  it('should render without the toggle icon', () => {
    const wrapper = shallow(
      <ListGroupItem hasSubItems={false}>foo</ListGroupItem>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.children().length).toBe(1);
  });

  it('should render with the toggle icon "-"', () => {
    const wrapper = shallow(
      <ListGroupItem hasSubItems isOpen level={1}>
        foo
      </ListGroupItem>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.children().length).toBe(2);
    expect(
      wrapper
        .childAt(0)
        .children()
        .text()
    ).toBe('-');
  });
  it('should render with the toggle icon "+"', () => {
    const wrapper = shallow(
      <ListGroupItem hasSubItems isOpen={false} level={1}>
        foo
      </ListGroupItem>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.children().length).toBe(2);
    expect(
      wrapper
        .childAt(0)
        .children()
        .text()
    ).toBe('+');
  });
  it('should render with level 3', () => {
    const wrapper = shallow(
      <ListGroupItem hasSubItems isOpen={false} level={3}>
        foo
      </ListGroupItem>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.prop('level')).toEqual(3);
    expect(wrapper.childAt(0).prop('level')).toEqual(3);
  });
});
