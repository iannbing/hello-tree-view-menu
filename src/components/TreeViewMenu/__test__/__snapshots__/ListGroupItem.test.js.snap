// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeViewMenu should render with level 3 1`] = `
<Styled(ListGroupItem)
  level={3}
  tag="li"
>
  <Styled(div)
    level={3}
  >
    +
  </Styled(div)>
  foo
</Styled(ListGroupItem)>
`;

exports[`TreeViewMenu should render with the toggle icon "+" 1`] = `
<Styled(ListGroupItem)
  level={1}
  tag="li"
>
  <Styled(div)
    level={1}
  >
    +
  </Styled(div)>
  foo
</Styled(ListGroupItem)>
`;

exports[`TreeViewMenu should render with the toggle icon "-" 1`] = `
<Styled(ListGroupItem)
  level={1}
  tag="li"
>
  <Styled(div)
    level={1}
  >
    -
  </Styled(div)>
  foo
</Styled(ListGroupItem)>
`;

exports[`TreeViewMenu should render without the toggle icon 1`] = `
<Styled(ListGroupItem)
  level={0}
  tag="li"
>
  foo
</Styled(ListGroupItem)>
`;
